third_party/tlx/dialect/include/Analysis/LayoutPropagation.h

```diff
@@ -0,0 +1,102 @@
+#ifndef TLX_ANALYSIS_LAYOUTPROPAGATION_H
+#define TLX_ANALYSIS_LAYOUTPROPAGATION_H
+
+#include "mlir/Analysis/DataFlow/SparseAnalysis.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/Support/LLVM.h"
+#include "triton/Dialect/Triton/IR/Dialect.h"
+#include <optional>
+
+using namespace mlir::dataflow;
+
+namespace mlir::triton::tlx {
+
+//===----------------------------------------------------------------------===//
+// LayoutEncoding
+//===----------------------------------------------------------------------===//
+
+class LayoutEncoding {
+public:
+  /// Construct a LayoutEncoding value as uninitialized.
+  explicit LayoutEncoding() = default;
+
+  /// Construct a LayoutEncoding value with a known constant.
+  LayoutEncoding(Attribute encoding) : encoding(std::move(encoding)) {}
+
+  bool operator==(const LayoutEncoding &rhs) const {
+    return encoding == rhs.encoding;
+  }
+
+  /// Whether the state is uninitialized.
+  bool isUninitialized() const { return !encoding.has_value(); }
+
+  /// Whether the state is unknown.
+  bool isUnknown() const { return encoding == nullptr; }
+
+  Attribute getLayoutEncoding() const {
+    assert(!isUninitialized());
+    return *encoding;
+  }
+
+  void print(raw_ostream &os) const;
+  static LayoutEncoding meet(const LayoutEncoding &lhs,
+                             const LayoutEncoding &rhs);
+  static LayoutEncoding join(const LayoutEncoding &lhs,
+                             const LayoutEncoding &rhs);
+  static LayoutEncoding getUnknownLayout() {
+    return LayoutEncoding{/*layoutEncoding=*/nullptr};
+  }
+
+private:
+  std::optional<Attribute> encoding;
```

Comment:
As discussed offline, can we add a flag indicating the encoding is a strong encoding (required by users/FE) or a weak encoding (inherits from a default encoding), and process the flag during meet/join? It could be a separate PR.